tket/src/Placement/NeighbourPlacements.cpp

```diff
@@ -0,0 +1,105 @@
+#include "NeighbourPlacements.hpp"
+
+#include <cstdlib>
+
+#include "TokenSwapping/SwapListOptimiser.hpp"
+
+namespace tket {
+
+NeighbourPlacements::NeighbourPlacements(
+    const Architecture& arc, const qubit_mapping_t& init_map)
+    : arc_(arc), init_map_(init_map), u_to_node_(), rng_() {
+  auto nodes = arc_.get_all_nodes_vec();
+  for (unsigned i = 0; i < nodes.size(); ++i) {
+    u_to_node_.left.insert({i, nodes[i]});
+  }
+}
+
+NeighbourPlacements::ResultVec NeighbourPlacements::get(
+    unsigned dist, unsigned n, bool optimise, unsigned seed) {
+  rng_.set_seed(seed);
+  ResultVec res;
+  for (unsigned i = 0; i < n; ++i) {
+    res.push_back(gen_result(dist, optimise));
+  }
+  return res;
+}
+
+NeighbourPlacements::Result NeighbourPlacements::gen_result(
+    unsigned dist, bool optimise) {
+  SwapList swaps;
+  tsa_internal::SwapListOptimiser optimiser;
+
+  // it might be impossible to find `dist` non-trivial swaps
+  unsigned n_unsuccessful = 0;
+  const unsigned N_TRIES = 10;
+
+  while (swaps.size() < dist && n_unsuccessful < N_TRIES) {
+    unsigned d = dist - swaps.size();
+    SwapList new_swaps = gen_swap_list(d);
+
+    if (optimise) {
+      SwapList swaps_candidate = swaps;
+      for (auto swap : new_swaps.to_vector()) {
+        swaps_candidate.push_back(swap);
+      }
+      optimiser.full_optimise(swaps_candidate);
+      if (swaps_candidate.size() > swaps.size()) {
+        swaps = std::move(swaps_candidate);
+        n_unsuccessful = 0;
+      } else {
+        ++n_unsuccessful;
+      }
+    } else {
+      for (auto swap : new_swaps.to_vector()) {
+        swaps.push_back(swap);
+      }
+    }
+  }
+
+  if (n_unsuccessful == N_TRIES) {
```

Comment:
What would be a good backup result for this case? Would it be possible to just use the initial mapping given as input?

Reply:
You could return a placement with as many swaps (`< dist`) as possible. Would you prefer that?

Reply:
Do you know how much additional runtime would be needed for an "as many swaps (< dist) as possible" solution?

Reply:
Sorry, I meant: I can return whatever maximum number of swaps I was able to generate, ie until I cannot greedily append swaps any further (not actually try to find a maximum).

Reply:
So there would be no additional runtime? For me that sounds like a good solution to just return that?

Reply:
OK. In that case, I will also return less than `n` solutions when I am unable to generate `n`. (see lines 60-73 in `NeighbourPlacements.cpp`)